### DISCUSSION TITLE
[Question] [Defrag and Compaction] - Fundamental Understanding Question

### DISCUSSION THREAD

**Discussion by vivekpatani** [etcd-io/etcd#17310] [Q&A] [answered]
Say I had a cluster that's been in production for a while, compaction occurs every 72h, but I've never defragged the cluster. Does that mean the tombstoned keys will continue to occupy space until defragged?

Background: Database has only 350ish keys
The size of the db reported is: 8 Gigs
Inspected the db and has a lot more than 350 keys.

Just wanted to confirm my theory, thank you.

**Comment by jmhbnz** [comment, marked as answer]
> Does that mean the tombstoned keys will continue to occupy space until defragged?

Short answer yes.

Compacting old revisions internally fragments etcd by leaving gaps in backend database. Fragmented space is available for use by etcd but unavailable to the host filesystem. In other words, deleting application data does not reclaim the space on disk.
- Ref: https://etcd.io/docs/v3.5/op-guide/maintenance/#defragmentation

If you use etcd with a space quota (which you should) then over time fragmentation will mean etcd will run out of storage space.
- Ref: https://etcd.io/docs/v3.5/op-guide/maintenance/#space-quota

**Comment by vivekpatani** [comment]
Thank you.
